core/src/hir/type_context.rs

```diff
@@ -464,51 +465,69 @@ impl TypeContext {
         }
     }
 
-    fn validate_primitive_slice_struct<P: super::TyPosition>(
-        &self,
-        errors: &mut ErrorStore,
-        st: &P::StructPath,
-    ) {
-        let ty = self.resolve_type(st.id());
-        match ty {
-            TypeDef::Struct(st) => {
-                if !st.attrs.allowed_in_slices {
-                    errors.push(LoweringError::Other(format!(
-                        "Cannot construct a slice of {:?}. Try marking with `#[diplomat::attr(auto, allowed_in_slices)]`",
-                        st.name
-                    )));
-                    return;
-                }
+    fn validate_type_def(&self, errors: &mut ErrorStore, def: TypeDef) {
+        if let TypeDef::Struct(st) = def {
+            self.validate_struct(errors, st);
```

Comment:
nit: probably should also have a `validate_trait` so that we can `validate_ty` everything in traits as well